File: common/src/main/java/me/pandamods/pandalib/config/holders/ClientConfigHolder.java
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package me.pandamods.pandalib.config.holders;

import me.pandamods.pandalib.config.Config;
import me.pandamods.pandalib.config.ConfigData;
import me.pandamods.pandalib.platform.Services;
import me.pandamods.pandalib.utils.Env;
import net.minecraft.world.entity.player.Player;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class ClientConfigHolder<T extends ConfigData> extends ConfigHolder<T> {
	private final Map<UUID, T> configs = new HashMap<>();

	public ClientConfigHolder(Class<T> configClass, Config config) {
		super(configClass, config);
	}

	@Override
	public void save() {
		if (Services.GAME.getEnvironment().equals(Env.CLIENT)) {
			super.save();
		} else
			this.logger.warn("Client config '{}' can't be saved on server", this.getDefinition().name());
	}

	@Override
	public boolean load() {
		if (Services.GAME.getEnvironment().equals(Env.CLIENT)) {
			return super.load();
		}
		return false;
	}

	@SuppressWarnings("unchecked")
	public <C extends ConfigData> void putConfig(Player player, C config) {
		configs.put(player.getUUID(), (T) config);
	}

	public T getConfig(Player player) {
		if (configs.containsKey(player.getUUID()))
			return configs.get(player.getUUID());
		return this.get();
	}
}

File: common/src/main/kotlin/me/pandamods/pandalib/config/holders/ClientConfigHolder.kt
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package me.pandamods.pandalib.config.holders

import dev.architectury.platform.Platform
import dev.architectury.utils.Env
import me.pandamods.pandalib.config.Config
import me.pandamods.pandalib.config.ConfigData
import net.minecraft.world.entity.player.Player
import java.util.*

class ClientConfigHolder<T : ConfigData>(configClass: Class<T>, config: Config) : ConfigHolder<T>(configClass, config) {
	private val configs: MutableMap<UUID?, T> = HashMap<UUID?, T>()

	public override fun save() {
		if (Platform.getEnvironment() == Env.CLIENT) {
			super.save()
		} else this.logger.warn("Client config '{}' can't be saved on server", this.definition.name)
	}

	public override fun load(): Boolean {
		if (Platform.getEnvironment() == Env.CLIENT) {
			return super.load()
		}
		return false
	}

	fun <C : ConfigData?> putConfig(player: Player, config: C?) {
		configs.put(player.getUUID(), config as T)
	}

	fun getConfig(player: Player): T? {
		if (configs.containsKey(player.getUUID())) return configs[player.getUUID()]
		return this.get()
	}
}
